e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java

```diff
@@ -65,4 +70,57 @@ public CipherWithNonce encryptBox(byte[] message, byte[] publicKey, byte[] priva
   public byte[] newSecretKey() {
     return random.randomBytes(CRYPTO_SECRETBOX_XSALSA20POLY1305_KEYBYTES);
   }
+
+  @Override
+  public byte[] signature(byte [] message, byte[] signingKey) {
+    checkNotNull(message, "message");
+    checkNotNull(signingKey, "signingKey");
+
+    byte[] combined = new byte[CRYPTO_SIGN_ED25519_BYTES + message.length];
+    int result = NaCl.sodium().crypto_sign_ed25519(combined, null, message, message.length, signingKey);
+
+    if(result != 0)
+      throw new RuntimeException("crypto_sign_ed25519: " + result);
+
+    return Arrays.copyOf(combined, CRYPTO_SIGN_ED25519_BYTES);
+  }
+
+  @Override
+  public boolean verify(Signature signature, byte[] message, byte[] publicSigningKey) {
+    checkNotNull(signature, "bytes");
+    checkNotNull(message, "message");
+    checkNotNull(publicSigningKey, "publicSigningKey");
+
+    byte[] combined = new byte[signature.bytes.length + message.length];
+    System.arraycopy(signature.bytes, 0, combined, 0, signature.bytes.length);
+    System.arraycopy(message, 0, combined, signature.bytes.length, message.length);
+    return NaCl.sodium().crypto_sign_ed25519_open(new byte[message.length], null, combined, combined.length, publicSigningKey) == 0;
```

Comment:
I think you can go higher-level with at least this one:
```java
return new VerifyKey(publicSigningKey).verify(message, signature.bytes)
```

Reply:
That throws a RuntimeException when verification fails. Can use `try`/`catch` of course, but doesn't seem worth it.

Reply:
Ugh, gross.